### PR TITLE
AIP-72: Handling skipped tasks in task_sdk

### PR DESCRIPTION
**Pull request by amoghrajesh**
related: https://github.com/apache/airflow/issues/44414

We already have support for handling terminal states from the task execution side as well as the task SDK client side. (almost).

This PR extends the task runner's `run` function to handle when an AirflowSkipException is thrown.

1. Extended the task runner with the right `msg` when a AirflowSkipException is thrown
2. Added a test in the `test_handle_requests` test of supervisor to show that no `client` call is made from `handle_requests` but is instead controlled by the `wait` function.
3. Added an example DAG that throws AirflowSkipException and the task runner test that handles that.

Postman working example:
1. Create a DAG run:
image

2. Login into metadata DB and get the task_id for the task you want to skip
image

3. Once you have the task_id, send a patch request like this:
```
curl --location --request PATCH 'http://localhost:29091/execution/task-instances/0193ab1b-df6a-72bb-b888-04c541e6bf12/state' \
--data '{
    "state": "skipped",
    "end_date": "2024-10-31T12:00:00Z"
}'
```

image

4. Check the Airflow UI if the task was marked as "skipped"
image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: task_sdk/tests/dags/super_basic_skipped.py
Comment (ashb) [inline task_sdk/tests/dags/super_basic_skipped.py:1]: Let's not keep the "super_basic" prefix unless it does "nothing" at all advanced.


I also wonder if it's worth building up the mock/fake capability to define the DAG inline in the test and inject it, for example maybe we could stub out the "parse" function using KGB's `spy_agency` to make it use the DAG we want (I'm just somewhat wary of an ever-growing `tests/dags/` folder, as ti makes it harder to know if those dags are still in use.)

Comment (amoghrajesh) [reply task_sdk/tests/dags/super_basic_skipped.py:1]: Made it basic_* instead.

Comment (amoghrajesh) [reply task_sdk/tests/dags/super_basic_skipped.py:1]: @ashb are you OK with me picking the second part of stubbing the parse and related things in a follow up? It isn't directly related to this PR

Comment (ashb) [reply task_sdk/tests/dags/super_basic_skipped.py:1]: Yeah sounds good!

File: task_sdk/tests/execution_time/test_supervisor.py [hunk 1]
```diff
@@ -883,7 +892,8 @@ def test_handle_requests(
         generator.send(msg)
 
         # Verify the correct client method was called
-        mock_client_method.assert_called_once_with(*method_arg)
+        if client_attr_path:
```
Comment (ashb) [inline task_sdk/tests/execution_time/test_supervisor.py:895]: How does L884 work when client_attr_path is empty?

File: task_sdk/tests/execution_time/test_task_runner.py [hunk 2]
```diff
@@ -137,3 +137,27 @@ def test_run_deferred_basic(test_dags_dir: Path, time_machine):
...
+
+    ti = parse(what)
+
+    instant = timezone.datetime(2024, 12, 3, 10, 0)
+    time_machine.move_to(instant, tick=False)
+
+    with mock.patch(
+        "airflow.sdk.execution_time.task_runner.SUPERVISOR_COMMS", create=True
+    ) as mock_supervisor_comms:
+        mock_supervisor_comms.send_request = mock.Mock()
```
Comment (ashb) [inline task_sdk/tests/execution_time/test_task_runner.py]: Nit: I don't think this is needed -- won't mock handle this for us automatically?

File: task_sdk/tests/execution_time/test_supervisor.py (hunk 1 above)
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_supervisor.py:895]: It just goes ahead with a blank mock, I let it be that that way as it is only used to define a `mock_response` which we have defined above in parameterize.
image

File: task_sdk/tests/execution_time/test_task_runner.py (hunk 2 above)
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_task_runner.py]: Yeah, it should

Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_task_runner.py]: Handled in latest commit

**Review by ashb — APPROVED** [review]

**Comment by amoghrajesh** [comment]
This one is pretty straight forward. Merging it

**Pull request merged**